cabal-install/Distribution/Client/Init.hs

```diff
@@ -207,8 +207,22 @@ getLicense flags = do
          ?>> fmap (fmap (either UnknownLicense id))
                   (maybePrompt flags
                     (promptList "Please choose a license" listedLicenses (Just BSD3) display True))
-  return $ flags { license = maybeToFlag lic }
+
+  invalidOtherLicense <- if isLicenseInvalid lic
+                            then putStrLn promptInvalidOtherLicenseMsg >> return True
+                            else return False
+
+  if invalidOtherLicense
+    then getLicense flags
+    else return $ flags { license = maybeToFlag lic }
+
   where
+    isLicenseInvalid (Just (UnknownLicense t)) = any (not . isAlphaNum) t
+    isLicenseInvalid _ = False
+
+    promptInvalidOtherLicenseMsg = "\nThe license must be alphanumeric. " ++ 
```

Comment:
It might be worth noting that if a license name has many words in it, the convention is to camel case it (e.g., PublicDomain.) This gives the user better actionable advice.